### ISSUE TITLE
How to optimize the coarse SR network?

### ISSUE THREAD

**Issue by XiangyuWu** [tyshiwo/FSRNet#5]
Hi, Tai! I'm reading your paper and trying to reproduce your work recently. And my question is how to optimize the coarse SR network. Are the gradients coming from the end of the network? 

And in the beginning of the net training, how to set the initial weights of the 3*3 conv layer in coarse SR network to ensure that the images generated are of relatively good quality then the origin low-resolution images.

**Comment by XiangyuWu** [comment]
Seems that I got some misunderstandings. FSRNet should be regarded as a whole, and each part is trainable.

**Comment by Errorkiller** [comment]
同学，我最近也看了这篇论文，我们可以讨论下吗？我的QQ：626712060

**Comment by XiangyuWu** [comment]
Hi, ty!

I implemented your network recently, and the final output performed well. However, as the question mentioned before, the intermediate output from coarse SR network seemed a little bit strange.

Is it normal? It seems that the coarse SR network is hard to optimize for me.

![train](https://user-images.githubusercontent.com/38173114/46785257-d4c5f680-cd63-11e8-9731-937af91b5090.jpg)

**Comment by XiangyuWu** [comment]
Hhhhh, my fault again in my network structure.

**Comment by nmber5** [comment]
@XiangyuWu  hi xiangyu, can you share a copy of your fsrnet?

**Comment by XiangyuWu** [comment]
@nmber5 I‘m afraid not. But if you meet some problems in your work, any contact is welcomed.

**Comment by cydiachen** [comment]
> @nmber5 I‘m afraid not. But if you meet some problems in your work, any contact is welcomed.

Hello, I am trying to implement my own version of FSRNET, I meet with some of the problem in implement the Fine SR Encoder-Decoder part, Would you mind giving me some help? Thanks a lot

**Comment by cydiachen** [comment]
> @nmber5 I‘m afraid not. But if you meet some problem

> I implemented your network recently, and the final output performed well. However, as the question mentioned before, the intermediate output from coarse SR network seemed a little bit weird.

Would you mind giving me some of the advice on implementation of SR network, I am now meet up with some problems, You can contact me via QQ or wechat: 1017259898

**Comment by nmber5** [comment]
> @nmber5 I‘m afraid not. But if you meet some problems in your work, any contact is welcomed.

hi @XiangyuWu, i met some trouble in test these models.

firstly, i tried this model with the data given by @tyshiwo the results is very good.
secondly, when i use my own face img, the result is not satisfied.
               i just resize the face img to size of(128, 128) and then fed them to fsrnet, the result is very strange, can you help on how to generate train/test LR img?



my email is:   nmber5@live.com

thanks.

**Comment by XiangyuWu** [comment]
@nmber5 
For generating LR faces, _cv2.resize(image, (128,128), cv2.INTER_CUBIC)_ is what I used.

For the question why you cannot generate the satisfied results, different datasets follow different distributions. In our view, all inputs are faces in low resolution, but the distributions of faces in one dataset vary from that in another dataset. So you should train different models for different datasets.

**Comment by nmber5** [comment]
> @nmber5
> For generating LR faces, _cv2.resize(image, (128,128), cv2.INTER_CUBIC)_ is what I used.
> 
> For the question why you cannot generate the satisfied results, different datasets follow different distributions. In our view, all inputs are faces in low resolution, but the distributions of faces in one dataset vary from that in another dataset. So you should train different models for different datasets.

@XiangyuWu tks a lot, it seems a new model  should trained for my situation

**Comment by nmber5** [comment]
> @nmber5 I‘m afraid not. But if you meet some problems in your work, any contact is welcomed.

hi xiangyu, 
can you give me some help.
i trying to parse  the given model to re-implement my own net.
but i was messed with  the following 2-4 and 6-8 tensor. are they bias? i was confused: if they are bias, only one should be. but here repeated 3 times? 
  1 : CudaTensor - size: 64x3x3x3
  2 : CudaTensor - size: 64
  3 : CudaTensor - size: 64
  4 : CudaTensor - size: 64
  5 : CudaTensor - size: 32x64x1x1
  6 : CudaTensor - size: 32
  7 : CudaTensor - size: 32
  8 : CudaTensor - size: 32

**Comment by XiangyuWu** [comment]
> > @nmber5 I‘m afraid not. But if you meet some problems in your work, any contact is welcomed.
> 
> hi xiangyu,
> can you give me some help.
> i trying to parse the given model to re-implement my own net.
> but i was messed with the following 2-4 and 6-8 tensor. are they bias? i was confused: if they are bias, only one should be. but here repeated 3 times?
> 1 : CudaTensor - size: 64x3x3x3
> 2 : CudaTensor - size: 64
> 3 : CudaTensor - size: 64
> 4 : CudaTensor - size: 64
> 5 : CudaTensor - size: 32x64x1x1
> 6 : CudaTensor - size: 32
> 7 : CudaTensor - size: 32
> 8 : CudaTensor - size: 32

I am not clear about what's that. Maybe @tyshiwo would help you with it.

**Comment by XiaotianM** [comment]
Hi, I have a question. The coarse SR network has not upsampled the input, so how to optimize the output of coarse SR network with GT HR image?

**Comment by XiaotianM** [comment]
@XiangyuWu  @tyshiwo

**Comment by XiangyuWu** [comment]
> Hi, I have a question. The coarse SR network has not upsampled the input, so how to optimize the output of coarse SR network with GT HR image?

By gradients from fine SR network. Also, I think you can add supervision like MSE loss between coarse SR output and GT use MSE to guide the coarse SR network if you want to. In my view, coarse SR network is for prior estimation, since the prior estimation network is pretrained, which need inputs of rgb image. You may use embedding rather than coarse SR image for fine SR network.

**Comment by XiaotianM** [comment]
@XiangyuWu, Thanks for your reply. At first, I thought the input size is smaller than HR. So it can not calculate mse loss between two images with different sizes. Now I think I got it. The input is upsampled as GT size first.
